### PR TITLE
Adding Action Annotations

### LINKED ISSUE #609: Annotations on Actions only work if "web export" exists
It appears at some point (perhaps when we added support for "web actions" that the ComposeActions() function has logic that ONLY creates Annotations on the wskAction IFF ```action.Webexport == "true"```. 

To make matters more convoluted, even if "Webexport==true", the annotations are created as a "side effect" of calling ```utils.WebAction```; this is not good coding practice.

```		keyValArr = make(whisk.KeyValueArr, 0)
		for name, value := range action.Annotations {
			var keyVal whisk.KeyValue
			keyVal.Key = name
			keyVal.Value = utils.GetEnvVar(value)
			keyValArr = append(keyValArr, keyVal)
		}

		// only set the webaction when the annotations are not empty.
		if action.Webexport == "true" {
			// TODO() why is this commented out?  we should now support annotations...
			//wskaction.Annotations = keyValArr
			wskaction.Annotations, errorParser = utils.WebAction("yes", keyValArr, action.Name, false)
			if errorParser != nil {
				return s1, errorParser
			}
		}
```

we have NO testcases for Annotations on Actions (only for Triggers) and need several as part of fixing this issue.  The Annotations SHOULD work with or without "webExport==true" and NOT be a side effect.

### PR DESCRIPTION
**Pull request by pritidesai**
Closes #609 

_Please don't get scared with the number of files changed :) this number includes YAML files for unit tests and updates to YAML files of existing integration test and a new integration test directory._

Here is the list of changes:

- Moved web action code out of `misc.go` into `webaction.go`
- Populating action annotations irrespective of `web-export`
- Crucial changes when action annotations are specified in deployment file:
    - when a deployment file has a new action annotation which is not defined in manifest file, `wskdeploy` errors out saying `Error: wskdeploy.go [125]: [ERROR_COMMAND_FAILED]: wskdeploy: Error: deploymentreader.go [227]: [ERROR_YAML_FORMAT_ERROR]: Annotation key <key> does not exist in manifest file but specified in deployment file.
`
    - when a deployment file has defined annotation which is also defined in manifest file, `wskdeploy` overwrites annotation value from deployment i.e. annotation value in deployment file takes higher precedence over the value in manifest file.
- Added support for annotations on package and triggers with same precedence order as action annotations

### CODE REVIEW

File: deployers/deploymentreader.go [hunk 1]
```diff
@@ -121,25 +126,32 @@ func (reader *DeploymentReader) bindPackageInputsAndAnnotations() {
...
+						serviceDeployPack.Package.Annotations[i].Value = input
+						break
+					}
+				}
+				if !keyExistsInManifest {
+					return utils.NewYAMLFormatError("Annotation key \""+name+"\" does not exist in manifest file but specified in deployment file.")
+				}
 			}
-
-			serviceDeployPack.Package.Annotations = keyValArr
```
Comment (mrutkows) [inline deployers/deploymentreader.go]: Is the array always allocated now?

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go [hunk 2]
```diff
@@ -70,10 +76,10 @@ func TestBindingInputsAnnotations(t *testing.T) {
 	}
 	for _, annos := range wskaction.Annotations {
 		switch annos.Key {
-		case "tag":
-			assert.Equal(t, "hello", annos.Value, "Failed to get action annotations")
-		case "aaa":
-			assert.Equal(t, "this is an annotation", annos.Value, "Failed to get action annotations")
+		case "action_annotation_1":
+			assert.Equal(t, "this is a action annotation 1", annos.Value, "Failed to get action annotations")
```
Comment (mrutkows) [inline tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: c /a/an/

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go [hunk 3]
```diff
@@ -70,10 +76,10 @@ func TestBindingInputsAnnotations(t *testing.T) {
...
+		case "action_annotation_2":
+			assert.Equal(t, "this is a action annotation 2", annos.Value, "Failed to get action annotations")
```
Comment (mrutkows) [inline tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: c /a/an/

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go [hunk 4]
```diff
@@ -93,23 +99,58 @@ func TestBindingInputsAnnotations(t *testing.T) {
...
+		default:
+			assert.Fail(t, "Failed to get package annotation key")
+		}
+	}
+	// verify the annotations of action
+	wskaction = pkgobj.Actions["helloworld"].Action
+	for _, annos := range wskaction.Annotations {
+		switch annos.Key {
+		case "action_annotation_1":
+			assert.Equal(t, "this is a action annotation 1 from deployment", annos.Value, "Failed to get action annotations")
```
Comment (mrutkows) [inline tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: c /a/an/

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go [hunk 5]
```diff
@@ -93,23 +99,58 @@ func TestBindingInputsAnnotations(t *testing.T) {
...
+		case "action_annotation_2":
+			assert.Equal(t, "this is a action annotation 2", annos.Value, "Failed to get action annotations")
```
Comment (mrutkows) [inline tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: c /a/an/

File: tests/dat/manifest_validate_action_annotations.yaml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                        description: name of a person
+                    place:
+                        type: string
+                        description: location of a person
+                outputs:
+                    payload:
+                        type: string
+                        description: a simple greeting message, Hello World!
+                annotations:
+                    action_annotation_1: this is annotation 1
```
Comment (mrutkows) [inline tests/dat/manifest_validate_action_annotations.yaml]: Do we have a test for more complex (i.e., arrays/JSON) annotations?  Should we open another issue (feature) to support this?

File: deployers/deploymentreader.go [hunk 7]
```diff
@@ -121,25 +126,32 @@ func (reader *DeploymentReader) bindPackageInputsAndAnnotations() {
...
+					if name == a.Key {
+						// annotation key is found in manifest
+						keyExistsInManifest = true
+						// overwrite annotation in manifest file with deployment file
+						serviceDeployPack.Package.Annotations[i].Value = input
+						break
+					}
+				}
+				if !keyExistsInManifest {
+					return utils.NewYAMLFormatError("Annotation key \""+name+"\" does not exist in manifest file but specified in deployment file.")
```
[the remark this answers is not in the document]
Comment (pritidesai) [reply deployers/deploymentreader.go]: Done

File: deployers/deploymentreader.go (hunk 1 above)
Comment (pritidesai) [reply deployers/deploymentreader.go]: Yes if one or more annotations are specified ...

File: tests/dat/manifest_validate_action_annotations.yaml (hunk 6 above)
Comment (pritidesai) [reply tests/dat/manifest_validate_action_annotations.yaml]: Nope, created new issue https://github.com/apache/incubator-openwhisk-wskdeploy/issues/634

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go (hunk 2 above)
Comment (pritidesai) [reply tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: Done

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go (hunk 3 above)
Comment (pritidesai) [reply tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: Done

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go (hunk 4 above)
Comment (pritidesai) [reply tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: Done

File: tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go (hunk 5 above)
Comment (pritidesai) [reply tests/src/integration/validate-binding-inputs-annotations/validate-bind-inputs-anno_test.go]: Done

**Pull request merged**